### PR TITLE
support local storage fireproofing

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/1204888209490842/f
Tech Design URL:
CC: @samsymons, @bwaresiak 

**Description**:

Revert to using default persistence container for website data storage.

Also various refactoring to make future singleton slaying quests a bit easier.

**Steps to test this PR**:
1. Using an older version of the app (any before this PR should be fine).  Open some tabs, sign into to some websites, and fire proof them.  Check the fire proofing is working by using the fire button and leave the tabs open to simulate users receiving this update.  Then update the app to this branch. 
2. Use the fire button and ensure previously signed in sites still work as before.
3. Use the Fireproofing UI in settings to remove fireproofed sites and use the fire button to clear them, ensuring their data has been cleared.
4. Use the Cookies UI in debug to check there's no cookies.
5. Navigate to https://privacy-test-pages.site/features/local-storage.html and start the counter.  
6. Use the fire button and navigate back to https://privacy-test-pages.site/features/local-storage.html - previous values should be removed.
7. Fireproof the site and start the counter.  Wait for a while, then use the fire button. Navigate back to https://privacy-test-pages.site/features/local-storage.html and the previous values should be set.
8. Remove fireproofing for the site and use the fire button.  Navigate to it once more and ensure the values are again unset.
9. Do similar smoke testing using the fire button on the tab switcher
10. Do so similar smoke testing using auto-clear setting on
11. Run the `release_tests/local-storage` UI test and validate it works

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Device Testing**:

* [ ] iPhone
* [ ] iPad (smoke testing)

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16 (option)
* [ ] iOS 17 (option)
* [ ] iOS 18

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/WebCacheManager.swift [hunk 1]
```diff
@@ -34,26 +33,75 @@ extension WKWebsiteDataStore {
...
-    ///  the container only persists cookies to disk when the web view is used.  If the user presses the fire button
-    ///  twice then the fire proofed cookies will be lost and the user will be logged out any sites they're logged in to.
-    public func consumeCookies(cookieStorage: CookieStorage = CookieStorage(),
-                               httpCookieStore: WKHTTPCookieStore) async {
+    /// The previous version saved cookies externally to the data so we can move them between containers.  We now use
+    /// the default persistence so this only needs to happen once when the fire button is pressed.
+    ///
+    /// The migration code removes the key that is used to check for the isConsumed flag so will only be
+    ///  true if the data needs to be migrated.
+    public func consumeCookies(intoHTTPCookieStore httpCookieStore: WKHTTPCookieStore) async {
```
Comment (jaceklyp) [inline Core/WebCacheManager.swift]: httpCookieStore is strongly typed, so I believe it should be consumeCookies(into httpCookieStore: WKHTTPCookieStore) async {

File: Core/DataStoreIdManager.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+//  limitations under the License.
+//
+
+
+import WebKit
+import Persistence
+
+/// Supports an existing ID set in previous versions of the app, but moving forward does not allocate an ID.  We have gone back to using the default
+///  peristence for the webview storage so that we can fireproof types that don't have an API for accessing their data. (e.g. localStorage)
+public protocol DataStoreIdManaging {
```
Comment (jaceklyp) [inline Core/DataStoreIdManager.swift]: nitpick; I think we should be using ID (not Id) just like Apple uses it in their libs, e.g. UUID or https://developer.apple.com/documentation/swiftui/scrollposition/viewid(type:)

File: Core/Fireproofing.swift [hunk 3]
```diff
@@ -51,15 +54,19 @@ public class UserDefaultsFireproofing: Fireproofing {
         }
     }
 
+    var allowedDomainsIncludingDuckDuckGo: [String] {
```
Comment (jaceklyp) [inline Core/Fireproofing.swift]: private

File: Core/HTTPCookieExtension.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import WebKit
+
+extension HTTPCookie {
+
+    static func cookieDomain(_ cookieDomain: String, matchesTestDomain testDomain: String) -> Bool {
+        return testDomain == cookieDomain
+            || ".\(testDomain)" == cookieDomain
+            || (cookieDomain.hasPrefix(".") && testDomain.hasSuffix(cookieDomain))
+    }
```
Comment (jaceklyp) [inline Core/HTTPCookieExtension.swift:33]: could you provide a piece of documentation what do we test here for and possible e.g of what matches means for us

File: Core/MigratableCookieStorage.swift [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    /// Called after consuming cookies but migration has not happened.
+    /// This can happen in the following flow
+    /// * User presses fire button on container based app and then backgrounds it/switches to another app
+    /// * Installs update to this version of the app without using the browser
+    /// * Starts browsing
+    func setConsumed() {
+        resetStore()
+    }
```
Comment (jaceklyp) [inline Core/MigratableCookieStorage.swift:74]: im a bit confused, should anything else happen here besides "resetStore"?

File: Core/WebCacheManager.swift [hunk 6]
```diff
@@ -62,177 +110,102 @@ public class WebCacheManager {
...
-        var dataStore: WKWebsiteDataStore? = WKWebsiteDataStore(forIdentifier: containerId)
-        let cookies = await dataStore?.httpCookieStore.allCookies() ?? []
-        dataStore = nil
+    private func performMigrationIfNeeded(dataStoreIdManager: DataStoreIdManaging,
+                                          cookieStorage: MigratableCookieStorage,
+                                          destinationStore: WKWebsiteDataStore) async {
 
-        var uuids = await WKWebsiteDataStore.allDataStoreIdentifiers
+        // Check version here rather than on function so that we don't need complicated logic related to verison in the calling function
+        guard #available(iOS 17, *) else { return }
```
Comment (jaceklyp) [inline Core/WebCacheManager.swift:148]: why do we need this check?

**Review by jaceklyp — CHANGES_REQUESTED** [review]

File: Core/MigratableCookieStorage.swift (hunk 5 above)
Comment (brindy) [reply Core/MigratableCookieStorage.swift:74]: No, just wanted an API that made sense from where it was used. This is "behinds the scenes" behaviour related to migration.

File: Core/WebCacheManager.swift (hunk 6 above)
Comment (brindy) [reply Core/WebCacheManager.swift:148]: Minor optimisation - < iOS 17 no migration is needed. Will add comment.

File: Core/HTTPCookieExtension.swift (hunk 4 above)
Comment (brindy) [reply Core/HTTPCookieExtension.swift:33]: added comment and note to look at some unit tests

**Review by jaceklyp — APPROVED** [review]
LGTM :)

**Pull request merged**